File: 3362-zero-array-transformation-iii/3362-zero-array-transformation-iii.cpp
class Solution {
public:
    int maxRemoval(vector<int>& nums, vector<vector<int>>& queries) {
        int n=queries.size();
        int m = nums.size();
        sort(queries.begin(),queries.end());
        priority_queue<pair<int,int>> pq;
        vector<int> prefix(m+1);
        int i=0;  // nums index
        int j=0; // queries index
        while(i<m) {
            if(i>0){
                prefix[i]+=prefix[i-1];
            }

            // Load the queries covering index i 
            while(j<n && queries[j][0]<=i && i<=queries[j][1]) {
                pq.push({queries[j][1],queries[j][0]});
                j++;
            }

            while(!pq.empty() && nums[i]+prefix[i]>0 && pq.top().first>=i) {
                auto [u,l]= pq.top();
                prefix[i]-=1;
                prefix[u+1]+=1;
                pq.pop();
            }
            
            if(prefix[i]+nums[i]>0){
                return -1;
            }

            i++;
        }

        return pq.size();
    }
};


/*

0   1   2   3   4   5   6
1   2   3   4   5   6   7
-1  0   0   0   0   0   +1

-1  -1  -1  -1  -1  -1  0


queries= [0,5],[0,2]


[[1,3],[0,2],[1,3],[1,2]]

0   2
1   3
1   3
1   2



1. Sort the queries
2. Iterate nums
3. for each nums index i => load all the queries which cover index i into priority queue (such that the upper bound is in descending order)
4. update the query on prefix array until either the pq is exhausted or the nums[i]+prefix[i]==0
5. if pq exhausted and still this is not true => nums[i]+prefix[i]==0 then return -1
6. at the end return pq.size();

*/

File: 3362-zero-array-transformation-iii/3362-zero-array-transformation-iii.java
class Solution {
    public int maxRemoval(int[] nums, int[][] queries) {
        Arrays.sort(queries, new Comparator<int[]>() {
            @Override
            public int compare(int[] a, int[] b) {
                if(a[0]==b[0]) return Integer.compare(b[1],a[1]);
                return Integer.compare(a[0], b[0]); 
            }
        });
        PriorityQueue<Integer> choosenHeap = new PriorityQueue(); 
        PriorityQueue<Integer> candidateHeap = new PriorityQueue(new Comparator<Integer>() {
            @Override
            public int compare(Integer lhs, Integer rhs) {
                return rhs-lhs;
            }
        });
        int choosen = 0;
        int j = 0;
        for(int i=0;i<nums.length;i++) {
            // remove all invalid elements from the candidatesHeap
            while(!candidateHeap.isEmpty() && candidateHeap.peek()<i) candidateHeap.poll();

            // add all valid queries
            while(j<queries.length && queries[j][0]<=i && i<=queries[j][1]) {
                candidateHeap.add(queries[j][1]);
                j++;
            }
            // remove all invalid elements from the choosen heap
            while(!choosenHeap.isEmpty() && choosenHeap.peek()<i) choosenHeap.poll();  

            int reducedVal = nums[i]-choosenHeap.size();
            
            while(reducedVal>0 && !candidateHeap.isEmpty() && candidateHeap.peek()>=i) {
                choosenHeap.add(candidateHeap.poll());
                reducedVal -= 1;
                choosen++;
            }

            if(reducedVal>0) return -1; 
        }

        return queries.length-choosen;
    }
}


/*

Steps

1. Sort the intervals and initialize the candidates max heap and choosen min heap
2. iterate the nums array
3. for each nums[i]. Add the valid candidate's ending index into the maxHeap before adding remove the indexes which are no longer valid
4. validate the choosen heap and then check if the size of chosen heap is >= the value nums[i] if not then try to see if it happens by popping the candidates array into the choosen if not possible then return -1
6. keep count of choosen candidates 
5. repeat for all elements
7. return total queries - choosen count


*/